File: androidApp/src/main/kotlin/com/github/jvsena42/loopky/ui/signup/PhoneVerificationScreen.kt
package com.github.jvsena42.loopky.ui.signup

import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.OutlinedTextFieldDefaults
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.rememberUpdatedState
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.github.jvsena42.loopky.R
import com.github.jvsena42.loopky.presentation.signup.DialingCountry
import com.github.jvsena42.loopky.presentation.signup.PhoneVerificationEffect
import com.github.jvsena42.loopky.presentation.signup.PhoneVerificationPhase
import com.github.jvsena42.loopky.presentation.signup.PhoneVerificationUiState
import com.github.jvsena42.loopky.presentation.signup.PhoneVerificationViewModel
import com.github.jvsena42.loopky.ui.components.LoopkyPrimaryButton
import com.github.jvsena42.loopky.ui.theme.LoopkyTheme
import kotlinx.coroutines.flow.collectLatest
import org.koin.compose.viewmodel.koinViewModel

@Composable
fun PhoneVerificationRoute(onBack: () -> Unit, onDone: () -> Unit) {
    val viewModel: PhoneVerificationViewModel = koinViewModel()
    val state by viewModel.state.collectAsStateWithLifecycle()
    val currentOnDone by rememberUpdatedState(onDone)

    LaunchedEffect(viewModel) {
        viewModel.effects.collectLatest { effect ->
            when (effect) {
                PhoneVerificationEffect.NavigateToHandoff -> currentOnDone()
            }
        }
    }

    PhoneVerificationScreen(
        state = state,
        // Back from the code field returns to the number, keeping what was typed — the reason
        // both phases share one ViewModel.
        onBack = { if (state.phase == PhoneVerificationPhase.CodeEntry) viewModel.onBackToNumber() else onBack() },
        onPhoneChange = viewModel::onPhoneNumberChange,
        onCountrySelected = viewModel::onCountrySelected,
        onCodeChange = viewModel::onCodeChange,
        onSendCode = viewModel::onSendCodeClick,
        onVerify = viewModel::onVerifyClick,
    )
}

@Composable
private fun PhoneVerificationScreen(
    state: PhoneVerificationUiState,
    onBack: () -> Unit,
    onPhoneChange: (String) -> Unit,
    onCountrySelected: (DialingCountry) -> Unit,
    onCodeChange: (String) -> Unit,
    onSendCode: () -> Unit,
    onVerify: () -> Unit,
) {
    val isCodePhase = state.phase == PhoneVerificationPhase.CodeEntry
    SignupScaffold(
        title = stringResource(if (isCodePhase) R.string.signup_code_title else R.string.signup_phone_title),
        subtitle = if (isCodePhase) {
            stringResource(R.string.signup_code_subtitle, state.displayNumber)
        } else {
            stringResource(R.string.signup_phone_subtitle)
        },
        onBack = onBack,
        error = state.error,
    ) {
        // Split by phase rather than branching field-by-field: the two pages share a frame, not a
        // body, and threading `isCodePhase` through every property is how they become unreadable.
        if (isCodePhase) {
            CodeEntry(state = state, onCodeChange = onCodeChange, onVerify = onVerify, onResend = onSendCode)
        } else {
            NumberEntry(
                state = state,
                onPhoneChange = onPhoneChange,
                onCountrySelected = onCountrySelected,
                onSendCode = onSendCode,
            )
        }
    }
}

@Composable
private fun NumberEntry(
    state: PhoneVerificationUiState,
    onPhoneChange: (String) -> Unit,
    onCountrySelected: (DialingCountry) -> Unit,
    onSendCode: () -> Unit,
) {
    val colors = LoopkyTheme.colors
    var isPickingCountry by rememberSaveable { mutableStateOf(false) }
    FieldLabel(stringResource(R.string.signup_phone_label))
    SignupTextField(
        value = state.phoneNumber,
        onValueChange = onPhoneChange,
        isError = state.error != null,
        keyboardType = KeyboardType.Phone,
        testTag = "signup_phone_input",
        leading = {
            CountryCodeButton(
                country = state.country,
                enabled = !state.isWorking,
                onClick = { isPickingCountry = true },
                modifier = Modifier.testTag("signup_phone_country"),
            )
        },
    )
    Spacer(Modifier.height(8.dp))
    // Once the number is complete, read it back in full: the trunk `0` it may have dropped and the
    // calling code it added are both invisible in the field, and an SMS attempt is scarce.
    Text(
        text = if (state.canSendCode) {
            stringResource(R.string.signup_phone_sending_to, state.displayNumber)
        } else {
            stringResource(R.string.signup_phone_hint)
        },
        color = colors.foregroundMuted,
        fontSize = 12.sp,
        lineHeight = 16.sp,
        modifier = Modifier.testTag("signup_phone_hint"),
    )
    if (isPickingCountry) {
        CountryPickerSheet(
            selected = state.country,
            onSelect = {
                onCountrySelected(it)
                isPickingCountry = false
            },
            onDismiss = { isPickingCountry = false },
        )
    }
    Spacer(Modifier.height(24.dp))
    // Withheld on a terminal limit: offering "send" would invite the user to spend attempts they
    // no longer have.
    if (!state.isTerminal) {
        LoopkyPrimaryButton(
            label = stringResource(R.string.signup_phone_send),
            onClick = onSendCode,
            enabled = state.canSendCode,
            loading = state.isWorking,
            modifier = Modifier.testTag("signup_phone_send"),
        )
    }
}

@Composable
private fun CodeEntry(
    state: PhoneVerificationUiState,
    onCodeChange: (String) -> Unit,
    onVerify: () -> Unit,
    onResend: () -> Unit,
) {
    val colors = LoopkyTheme.colors
    FieldLabel(stringResource(R.string.signup_code_label))
    SignupTextField(
        value = state.code,
        onValueChange = onCodeChange,
        isError = state.error != null,
        keyboardType = KeyboardType.NumberPassword,
        testTag = "signup_code_input",
    )
    Spacer(Modifier.height(24.dp))
    if (!state.isTerminal) {
        LoopkyPrimaryButton(
            label = stringResource(R.string.signup_code_verify),
            onClick = onVerify,
            enabled = state.canVerify,
            loading = state.isWorking,
            modifier = Modifier.testTag("signup_code_verify"),
        )
        Spacer(Modifier.height(10.dp))
        TextButton(
            onClick = onResend,
            enabled = state.canResend,
            modifier = Modifier.testTag("signup_code_resend"),
        ) {
            Text(
                text = if (state.canResend) {
                    stringResource(R.string.signup_code_resend)
                } else {
                    stringResource(R.string.signup_code_resend_in, state.resendCooldownSeconds)
                },
                color = if (state.canResend) colors.accentSecondary else colors.foregroundMuted,
                fontSize = 13.sp,
            )
        }
    }
}

@Composable
private fun FieldLabel(text: String) {
    Text(text = text, color = LoopkyTheme.colors.foregroundMuted, fontSize = 11.sp)
    Spacer(Modifier.height(8.dp))
}

@Composable
private fun SignupTextField(
    value: String,
    onValueChange: (String) -> Unit,
    isError: Boolean,
    keyboardType: KeyboardType,
    testTag: String,
    leading: (@Composable () -> Unit)? = null,
) {
    val colors = LoopkyTheme.colors
    OutlinedTextField(
        value = value,
        onValueChange = onValueChange,
        modifier = Modifier.fillMaxWidth().testTag(testTag),
        leadingIcon = leading,
        singleLine = true,
        isError = isError,
        keyboardOptions = KeyboardOptions(keyboardType = keyboardType, autoCorrectEnabled = false),
        shape = RoundedCornerShape(12.dp),
        colors = OutlinedTextFieldDefaults.colors(
            focusedBorderColor = colors.accentPrimary,
            unfocusedBorderColor = colors.borderSubtle,
            cursorColor = colors.accentPrimary,
            errorBorderColor = colors.danger,
        ),
    )
}

File: iosApp/iosApp/Views/PhoneVerificationScreen.swift
import SwiftUI
import Shared

/// Verify by SMS. **One screen for both phases**, not two routes — splitting them would lose the
/// typed number when someone steps back to correct it.
struct PhoneVerificationScreen: View {
    var onBack: () -> Void
    var onDone: () -> Void

    @State private var viewModel: PhoneVerificationViewModel?
    @State private var uiState: PhoneVerificationUiState?
    @State private var stateSink: FlowEffectSink?
    @State private var effectSink: FlowEffectSink?
    @State private var phoneNumber = ""
    @State private var code = ""
    @State private var isPickingCountry = false

    private var isCodeEntry: Bool {
        uiState?.phase == PhoneVerificationPhase.codeentry
    }

    var body: some View {
        SignupScaffold(
            title: isCodeEntry ? "signup_code_title" : "signup_phone_title",
            subtitle: subtitle,
            errorTitle: SignupErrorCopy.title(for: uiState?.error),
            errorMessage: SignupErrorCopy.message(for: uiState?.error),
            // Back out of the code step returns to the number, keeping it — the same intent as
            // Android intercepting the system back gesture here.
            onBack: { isCodeEntry ? viewModel?.onBackToNumber() : onBack() }
        ) {
            if isCodeEntry { codeEntry } else { numberEntry }
        }
        .sheet(isPresented: $isPickingCountry) {
            if let uiState {
                CountryPickerSheet(selected: uiState.country) { viewModel?.onCountrySelected(country: $0) }
            }
        }
        .onAppear { attach() }
        .onDisappear { detach() }
    }

    private var subtitle: String {
        isCodeEntry
            ? String(format: NSLocalizedString("signup_code_subtitle", comment: ""),
                     uiState?.displayNumber ?? "")
            : NSLocalizedString("signup_phone_subtitle", comment: "")
    }

    private var numberEntry: some View {
        VStack(alignment: .leading, spacing: 0) {
            FieldLabel(text: "signup_phone_label")
            Spacer().frame(height: 8)
            HStack(spacing: 8) {
                CountryCodeButton(
                    country: uiState?.country ?? DialingCountries.shared.fallback,
                    isEnabled: !(uiState?.isWorking ?? false),
                    action: { isPickingCountry = true }
                )
                .accessibilityIdentifier("signup_phone_country")

                SignupTextField(
                    text: $phoneNumber,
                    placeholder: nil,
                    isEnabled: !(uiState?.isWorking ?? false),
                    isError: uiState?.error != nil,
                    keyboard: .phonePad
                )
                .textContentType(.telephoneNumber)
                .onChange(of: phoneNumber) { _, value in
                    viewModel?.onPhoneNumberChange(value: value)
                }
                .accessibilityIdentifier("signup_phone_input")
            }

            Spacer().frame(height: 8)
            // Once the number is complete, read it back in full: the trunk `0` it may have dropped
            // and the calling code it added are both invisible in the field.
            Text(verbatim: phoneHint)
                .font(.system(size: 12))
                .foregroundStyle(LoopkyColor.foregroundMuted)
                .fixedSize(horizontal: false, vertical: true)
                .accessibilityIdentifier("signup_phone_hint")

            // Withheld entirely on a terminal rate limit, not disabled: a button that cannot ever
            // work is worse than no button, because it invites tapping.
            if !(uiState?.isTerminal ?? false) {
                Spacer().frame(height: 24)
                SignupPrimaryButton(
                    title: "signup_phone_send",
                    isLoading: uiState?.isWorking ?? false,
                    isEnabled: uiState?.canSendCode ?? false,
                    action: { viewModel?.onSendCodeClick() }
                )
                .accessibilityIdentifier("signup_phone_send")
            }
        }
    }

    private var codeEntry: some View {
        VStack(alignment: .leading, spacing: 0) {
            FieldLabel(text: "signup_code_label")
            Spacer().frame(height: 8)
            SignupTextField(
                text: $code,
                placeholder: nil,
                isEnabled: !(uiState?.isWorking ?? false),
                isError: uiState?.error != nil,
                keyboard: .numberPad
            )
            .onChange(of: code) { _, value in viewModel?.onCodeChange(value: value) }
            .accessibilityIdentifier("signup_code_input")

            if !(uiState?.isTerminal ?? false) {
                Spacer().frame(height: 24)
                SignupPrimaryButton(
                    title: "signup_code_verify",
                    isLoading: uiState?.isWorking ?? false,
                    isEnabled: uiState?.canVerify ?? false,
                    action: { viewModel?.onVerifyClick() }
                )
                .accessibilityIdentifier("signup_code_verify")

                Spacer().frame(height: 10)
                Button(action: { viewModel?.onSendCodeClick() }) {
                    Text(verbatim: resendLabel)
                }
                .font(.system(size: 14, weight: .semibold))
                .foregroundStyle(canResend ? LoopkyColor.accentSecondary : LoopkyColor.foregroundMuted)
                .disabled(!canResend)
                .accessibilityIdentifier("signup_code_resend")
            }
        }
    }

    private var canResend: Bool { uiState?.canResend ?? false }

    private var phoneHint: String {
        guard let uiState, uiState.canSendCode else {
            return NSLocalizedString("signup_phone_hint", comment: "")
        }
        return String(format: NSLocalizedString("signup_phone_sending_to", comment: ""), uiState.displayNumber)
    }

    private var resendLabel: String {
        canResend
            ? NSLocalizedString("signup_code_resend", comment: "")
            : String(format: NSLocalizedString("signup_code_resend_in", comment: ""),
                     Int(uiState?.resendCooldownSeconds ?? 0))
    }

    private func attach() {
        guard viewModel == nil else { return }
        let vm = IosDependencies.shared.phoneVerificationViewModel()
        viewModel = vm
        stateSink = FlowEffectSink(vm.state) { value in
            guard let state = value as? PhoneVerificationUiState else { return }
            // The VM can restore a number on launch when a code was sent but never entered, so the
            // field follows state rather than only feeding it.
            // A pasted `+44 …` moves the picker and leaves only the national part, so the field
            // takes that back too. Only then: the field otherwise owns its own text while typing.
            if state.phoneNumber != phoneNumber && (phoneNumber.isEmpty || phoneNumber.contains("+")) {
                phoneNumber = state.phoneNumber
            }
            uiState = state
        }
        effectSink = FlowEffectSink(vm.effects) { effect in
            if effect is PhoneVerificationEffectNavigateToHandoff { onDone() }
        }
    }

    private func detach() {
        viewModel?.release()
        viewModel = nil
        stateSink = nil
        effectSink = nil
    }
}
